src/magic_enchantment.cpp

```diff
@@ -0,0 +1,453 @@
+            case enchantment::mod::ITEM_ATTACK_SPEED: return "ITEM_ATTACK_SPEED";
+            case enchantment::mod::ITEM_WET_PROTECTION: return "ITEM_WET_PROTECTION";
+            case enchantment::mod::NUM_MOD: break;
+        }
+        debugmsg( "Invalid enchantment::mod" );
+        abort();
+    }
+    // *INDENT-ON*
+
+template<>
+enchantment::has string_to_enum<enchantment::has>( const std::string &trigger )
+{
+    return string_to_enum_look_up( has_map, trigger );
+}
+template<>
+enchantment::condition string_to_enum<enchantment::condition>( const std::string &trigger )
+{
+    return string_to_enum_look_up( condition_map, trigger );
+}
+template<>
+enchantment::mod string_to_enum<enchantment::mod>( const std::string &trigger )
+{
+    return string_to_enum_look_up( mod_map, trigger );
+}
+} // namespace io
+
+namespace
+{
+generic_factory<enchantment> spell_factory( "enchantment" );
+} // namespace
+
+template<>
+const enchantment &string_id<enchantment>::obj() const
+{
+    return spell_factory.obj( *this );
+}
+
+template<>
+bool string_id<enchantment>::is_valid() const
+{
+    return spell_factory.is_valid( *this );
+}
+
+void enchantment::load_enchantment( JsonObject &jo, const std::string &src )
+{
+    spell_factory.load( jo, src );
+}
+
+bool enchantment::is_active( const Character &guy, const item &parent ) const
+{
+    if( !guy.has_item( parent ) ) {
+        return false;
+    }
+
+    if( active_conditions.first == has::HELD &&
+        active_conditions.second == condition::ALWAYS ) {
+        return true;
+    }
+
+    if( !( active_conditions.first == has::HELD ||
+           ( active_conditions.first == has::WIELD && &guy.weapon == &parent ) ||
+           ( active_conditions.first == has::WORN && guy.is_worn( parent ) ) ) ) {
+        return false;
+    }
+
+    if( active_conditions.second == condition::ALWAYS ) {
+        return true;
+    }
+
+    if( active_conditions.second == condition::UNDERGROUND ) {
+        if( guy.pos().z < 0 ) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    if( active_conditions.second == condition::UNDERWATER ) {
+        if( g->m.is_divable( guy.pos() ) ) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+    return false;
+}
+
+void enchantment::add_activation( const time_duration &dur, const fake_spell &fake )
+{
+    intermittent_activation[dur].emplace_back( fake );
+}
+
+void enchantment::load( JsonObject &jo, const std::string & )
+{
+    optional( jo, was_loaded, "id", id, enchantment_id( "" ) );
+    if( jo.has_array( "hit_you_effect" ) ) {
+        JsonArray jarray = jo.get_array( "hit_you_effect" );
+        while( jarray.has_more() ) {
+            fake_spell fake;
+            JsonObject fake_spell_obj = jarray.next_object();
+            fake.load( fake_spell_obj );
+            hit_you_effect.emplace_back( fake );
+        }
+    }
```

Comment:
I *think* you can replace this whole thing with `jo.read( "hit_you_effect", hit_you_effect );`.

Reply:
that works for arrays of objects?

Reply:
It should, yes.  See the test I added [here](https://github.com/jbytheway/Cataclysm-DDA/blob/621b4c19d98c443bc95f20ec74ed2a97e170ba2a/tests/json_test.cpp#L57-L58).